### PR TITLE
Remove java.util.Date

### PR DESCRIPTION
**Pull request by jianglai**
There is one remaining instance in JpaTransactionManagerImpl that cannot
be removed because DetachingTypedQuery is implementing TypedQuery, which has
a method that expectred java.util.Date.

- - -
This change is [Reviewable](https://reviewable.io/reviews/google/nomulus/2373)

### CODE REVIEW

**Review by jianglai — COMMENTED** [review]
> Reviewed 12 of 12 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2373)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @jianglai)

**Review by weiminyu — APPROVED** [review]
> Reviewed 12 of 12 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2373)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @jianglai)

File: config/presubmits.py [hunk 1]
```diff
@@ -173,17 +173,24 @@ def fails(self, file):
...
         "java",
         {"/node_modules/"},
     ):
         "Truth8 is deprecated. Use Truth instead.",
+    PresubmitCheck(
+        r".*java\.util\.Date.*",
+        "java",
+        {"/node_modules/", "JpaTransactionManagerImpl.java"},
+    ):
+        "Do not use java.util.Date. Use classes in Java.date package instead.",
```
Comment (CydeWeys) [inline config/presubmits.py:192]: What is the `Java.date` package?  That doesn't seem correct to me.

File: core/src/main/java/google/registry/flows/certs/CertificateChecker.java [hunk 2]
```diff
@@ -162,7 +158,7 @@ public ImmutableSet<CertificateViolation> checkCertificate(X509Certificate certi
     ImmutableSet.Builder<CertificateViolation> violations = new ImmutableSet.Builder<>();
 
     // Check if currently in validity period
-    Date now = clock.nowUtc().toDate();
+    DateTime now = clock.nowUtc();
```
Comment (CydeWeys) [inline core/src/main/java/google/registry/flows/certs/CertificateChecker.java:161]: This is a functionality change?  Are you sure this is even correct?

What's the motivation for removing `Date` everywhere anyway?  Sometimes you do legitimately need to process just a date, not a date and an attached time.

File: core/src/main/java/google/registry/tools/CreateOrUpdateBulkPricingPackageCommand.java [hunk 3]
```diff
@@ -56,9 +56,11 @@ abstract class CreateOrUpdateBulkPricingPackageCommand extends MutatingCommand {
   @Nullable
   @Parameter(
       names = "--next_billing_date",
+      converter = DateTimeParameter.class,
+      validateWith = DateTimeParameter.class,
       description =
           "The next date that the bulk pricing package should be billed for its annual fee")
-  Date nextBillingDate;
+  DateTime nextBillingDate;
```
Comment (CydeWeys) [inline core/src/main/java/google/registry/tools/CreateOrUpdateBulkPricingPackageCommand.java:63]: This seems like a bad change to me?  Why are we going to make people specify an entire date time on the command line when the only thing that should be passed in is a date?

File: networking/src/test/java/google/registry/networking/handler/SslInitializerTestUtils.java [hunk 4]
```diff
@@ -67,21 +66,26 @@ public static KeyPair getKeyPair() throws Exception {
...
     X500Name subjectDnName = new X500Name("CN=" + hostname);
     BigInteger serialNumber = BigInteger.valueOf(System.currentTimeMillis());
     X500Name issuerDnName = new X500Name(ssc.cert().getIssuerX500Principal().getName());
     ContentSigner sigGen = new JcaContentSignerBuilder("SHA256WithRSAEncryption").build(ssc.key());
     X509v3CertificateBuilder v3CertGen =
         new JcaX509v3CertificateBuilder(
-            issuerDnName, serialNumber, from, to, subjectDnName, keyPair.getPublic());
+            issuerDnName,
+            serialNumber,
+            from.toDate(),
```
Comment (CydeWeys) [inline networking/src/test/java/google/registry/networking/handler/SslInitializerTestUtils.java:85]: This is literally still using a `java.util.Date` then, only it's purposely being packaged into a DateTime, and then stripped out of it, for no reason!

**Review by jianglai — COMMENTED** [review]
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2373)* status: all files reviewed, 4 unresolved discussions (waiting on @CydeWeys)

___
*[`config/presubmits.py` line 192 at r2](https://reviewable.io/reviews/google/nomulus/2373#gh-1525338603:-NsyHN5F-kxf43TrOCme:brukr1b) ([raw file](https://github.com/google/nomulus/blob/c845f74cb70536fbcfa7225f6d36566edb26727a/config/presubmits.py#L192)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

What is the `Java.date` package?  That doesn't seem correct to me.
</blockquote></details>

My bad. Should be `java.time`.
___
*[`core/src/main/java/google/registry/flows/certs/CertificateChecker.java` line 161 at r2](https://reviewable.io/reviews/google/nomulus/2373#gh-1525340727:-NsyHzjt2S8KbhUfIgZd:b-c65d2m) ([raw file](https://github.com/google/nomulus/blob/c845f74cb70536fbcfa7225f6d36566edb26727a/core/src/main/java/google/registry/flows/certs/CertificateChecker.java#L161)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

This is a functionality change?  Are you sure this is even correct?

What's the motivation for removing `Date` everywhere anyway?  Sometimes you do legitimately need to process just a date, not a date and an attached time.
</blockquote></details>

`java.util.Date` is not just a date. It's [an instant in time with millisecond precision](https://docs.oracle.com/javase/8/docs/api/java/util/Date.html). There are many problems with it(no timezone support, no calendar support, etc) and most of its methods are deprecated for decades (since JDK 1.1). That's the whole reason why people prefer Joda Time instead of Java.util.Date.
___
*[`core/src/main/java/google/registry/tools/CreateOrUpdateBulkPricingPackageCommand.java` line 63 at r2](https://reviewable.io/reviews/google/nomulus/2373#gh-1525341435:-NsyJFRh86XoJ5T2-SDe:bcqkxmm) ([raw file](https://github.com/google/nomulus/blob/c845f74cb70536fbcfa7225f6d36566edb26727a/core/src/main/java/google/registry/tools/CreateOrUpdateBulkPricingPackageCommand.java#L63)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

This seems like a bad change to me?  Why are we going to make people specify an entire date time on the command line when the only thing that should be passed in is a date?
</blockquote></details>

Even with `Date` people are going to have to specify the time (see the other comment on `Data` being an instant in time). I don't know how jcommander handlers `Date` conversions, it may work if you don't specify one and it gives you some default time, but without timezone support that time is also ambiguous. We have been using `DateTime` in all other commands.

In general `java.util.Date` has such bad APIs that everyone should run from it unless they absolutely have to (like when their dependencies still expect it). Even then one should defer the conversion to `Date` until the last minute,  preferably manipulating Joda Time or `java.time` until they call `toDate()` to convert it and pass it to the dependencies.
___
*[`networking/src/test/java/google/registry/networking/handler/SslInitializerTestUtils.java` line 85 at r2](https://reviewable.io/reviews/google/nomulus/2373#gh-1525343457:-NsyGgg96900SvL6R20D:bwzcsa6) ([raw file](https://github.com/google/nomulus/blob/c845f74cb70536fbcfa7225f6d36566edb26727a/networking/src/test/java/google/registry/networking/handler/SslInitializerTestUtils.java#L85)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

This is literally still using a `java.util.Date` then, only it's purposely being packaged into a DateTime, and then stripped out of it, for no reason!
</blockquote></details>

We have no control over how our dependencies' interfaces are defined, but we can avoid directly interacting with the deprecated `java.util.date` API as much as possible.

**Review by jianglai — COMMENTED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2373)* status: all files reviewed, 4 unresolved discussions (waiting on @CydeWeys)

**Review by CydeWeys — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2373)* status: all files reviewed, 4 unresolved discussions (waiting on @jianglai)

___
*[`core/src/main/java/google/registry/flows/certs/CertificateChecker.java` line 161 at r2](https://reviewable.io/reviews/google/nomulus/2373#gh-1525340727:-Nsyj9jxDouWqszTzkFO:b-pq4eat) ([raw file](https://github.com/google/nomulus/blob/c845f74cb70536fbcfa7225f6d36566edb26727a/core/src/main/java/google/registry/flows/certs/CertificateChecker.java#L161)):*
<details><summary><i>Previously, jianglai (Lai Jiang) wrote…</i></summary><blockquote>

`java.util.Date` is not just a date. It's [an instant in time with millisecond precision](https://docs.oracle.com/javase/8/docs/api/java/util/Date.html). There are many problems with it(no timezone support, no calendar support, etc) and most of its methods are deprecated for decades (since JDK 1.1). That's the whole reason why people prefer Joda Time instead of Java.util.Date.
</blockquote></details>

How about https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/time/LocalDate.html then?  That looks to do what we need it to do here, but without that baggage.

**Review by jianglai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2373)* status: all files reviewed, 4 unresolved discussions (waiting on @CydeWeys)

___
*[`core/src/main/java/google/registry/flows/certs/CertificateChecker.java` line 161 at r2](https://reviewable.io/reviews/google/nomulus/2373#gh-1525340727:-NszDDjpB8Yfoj3dMznQ:b98ja26) ([raw file](https://github.com/google/nomulus/blob/c845f74cb70536fbcfa7225f6d36566edb26727a/core/src/main/java/google/registry/flows/certs/CertificateChecker.java#L161)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

How about https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/time/LocalDate.html then?  That looks to do what we need it to do here, but without that baggage.
</blockquote></details>

In this case, we definitely want both a date and a time (as it pertains to checking the expiration date of certificates or CRLs).

In general, the recommendation is to use `java.time` for greenfield projects and `joda.time` for existing project. We have a mixture of both at the moment, though I think it might be a good idea to pick a lane and stick to it to avoid unnecessary conversions. Joda actually recommends actively migrating to `java.time`, but we have a lot of `joda.time` and the migration might not be worth the effort.

**Review by jianglai — COMMENTED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2373)* status: all files reviewed, 4 unresolved discussions (waiting on @CydeWeys)

**Pull request merged**